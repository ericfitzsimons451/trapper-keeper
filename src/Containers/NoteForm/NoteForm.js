import React, { Component } from "react";
import { postNote } from "../../Thunks/postNote";
import { connect } from "react-redux";
import { editNote } from "../../Thunks/updateNote";

export class NoteForm extends Component {
  constructor() {
    super();
    this.state = {
      title: "",
      listItems: []
    };
  }

  componentDidMount() {
    if (this.props.note) {
      this.setState({
        title: this.props.note.title,
        listItems: this.props.note.listItems
      });
    }
  }

  addNote = e => {
    const { id } = e.target.parentElement;
    if (e.target.value !== null) {
      this.setState({
        listItems: [
          ...this.state.listItems,
          { id: parseInt(id), text: e.target.value, checked: false }
        ]
      });
    }
  };

  deleteNote = e => {
    const { id } = e.target.parentElement;
    const newBody = this.state.listItems.filter(note => {
      return note.id !== parseInt(id);
    });
    this.setState({
      listItems: newBody
    });
  };

  toggleCheckBox = e => {
    const { id } = e.target.parentElement;
    const newBody = this.state.listItems.map(note => {
      if (note.id === parseInt(id)) {
        return { text: note.text, checked: !note.checked, id: note.id };
      }
      return note;
    });
    this.setState({
      listItems: newBody
    });
  };

  editNoteText = e => {
    const { id } = e.target.parentElement;
    const { value } = e.target;
    const newBody = this.state.listItems.map(note => {
      if (note.id === parseInt(id)) {
        return { text: value, checked: note.checked, id: note.id };
      }
      return note;
    });
    this.setState({
      listItems: newBody
    });
  };

  editTitle = e => {
    const title = e.target.value;
    this.setState({
      title
    });
  };

  handleSubmit = async e => {
    e.preventDefault();
    if (this.props.note) {
      const updatedNote = {
        id: this.props.note.id,
        title: this.state.title,
        listItems: this.state.listItems
      };
      await this.props.editNote(updatedNote);
      this.props.history.push("/");
    } else {
      const { className, type } = e.target;
      if (className === "modal" || type === "submit") {
				await this.props.postNote(this.state);
        this.props.history.push("/");
      }
    }
  };

  render() {
    const { listItems } = this.state;
    const id = Date.now();
    const filteredUncheckedNotes = listItems.filter(note => !note.checked);
    const filteredCheckedNotes = listItems.filter(note => note.checked);
    const uncheckedNotes = filteredUncheckedNotes.map(text => {
      return (
        <div key={text.id} id={text.id} className="text">
          <div onClick={this.toggleCheckBox} className="uncheckbox" />
          <input
            placeholder="take a note"
            name="body"
            onChange={this.editNoteText}
            value={text.text}
          />
          <div onClick={this.deleteNote} className="xmark" />
        </div>
      );
    });
    const checkedNotes = filteredCheckedNotes.map(text => {
      return (
        <div key={text.id} id={text.id} className="text">
          <div onClick={this.toggleCheckBox} className="checkbox" />
          <input
            placeholder="take a note"
            name="body"
            onChange={this.editNoteText}
            value={text.text}
          />
          <div onClick={this.deleteNote} className="xmark" />
        </div>
      );
    });
    uncheckedNotes.push(
      <div key={id} id={id} className="text">
        <div className="add" />
        <input
          placeholder="take a note"
          name="body"
          onChange={this.addNote}
          value={""}
        />
      </div>
    );
    return (
      <div className="modal">
        <form className="note-form">
          <input className="title-input"
            onChange={this.editTitle}
            name="title"
            value={this.state.title}
            placeholder="title"
          />
          {uncheckedNotes}
          {checkedNotes}
          <nav>
            <button onClick={this.handleSubmit}>Save Note</button>
          </nav>
        </form>
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  postNote: note => dispatch(postNote(note)),
  editNote: note => dispatch(editNote(note))
});

export default connect(
  null,
  mapDispatchToProps
)(NoteForm);
