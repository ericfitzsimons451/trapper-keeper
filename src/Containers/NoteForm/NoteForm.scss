.modal {
    top: 0px;
    left: 0px;
    position: fixed;
    width: 100%;
    height: 100%;
    overflow: scroll;
    background: rgba($color: #000000, $alpha: 0.5);
    .title-input {
        border: none;
        outline: none;
        font-size: 2em;
        margin: 0px 12px;
    }
    .note-form {
        background: #ffffff;
        margin: 100px auto;
        width: 600px;
        min-height: 100px;
        padding: 20px 0px;
        box-shadow: 2px 2px 5px grey;
        border-radius: 5px;
        display: flex;
        flex-direction: column;
    }
}

.text {
    border-top: 1px solid #ffffff;
        border-bottom: 1px solid #ffffff;
    box-sizing: content-box;
    padding: 10px;
    display: flex;
    margin: 5px;
    input{
        width: 90%;
        margin: 0px 10px;
        &:-webkit-autofill,
        &:-webkit-autofill:hover, 
        &:-webkit-autofill:focus, 
        &:-webkit-autofill:active  {
            -webkit-box-shadow: 0 0 0 30px white inset !important;
        }
    }
    .xmark {
        cursor: pointer;
        transition: 0.2s;
        opacity: 0;
        height: 15px;
        width: 15px;
        background-image: url("./images/close-button.svg");
        background-size: 15px 15px;
        &:hover{
            .xmark {
                opacity: 1;
            } 
        }
        &:focus-within {
            border-top: 1px solid black;
            border-bottom: 1px solid black;
        }
    }
    input {
        border: none;
        outline: none;
    }
}

.uncheckbox {
    cursor: pointer;
    height: 15px;
    width: 15px;
    background-image: url("./images/blank-square.svg");
    background-size: 15px 15px;
}

.checkbox {
    cursor: pointer;
    height: 15px;
    width: 15px;
    background-image: url("./images/check-box.svg");
    background-size: 15px 15px;
}

.add {
    height: 15px;
    width: 15px;
    background-image: url("./images/add-plus-button.svg");
    background-size: 15px 15px;
}