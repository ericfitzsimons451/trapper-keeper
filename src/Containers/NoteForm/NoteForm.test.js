import React from "react";
import { shallow } from "enzyme";
import { NoteForm } from "./NoteForm";

describe.skip("NoteForm", () => {
  let wrapper;
  const mockFunc = jest.fn();
  const mockState = { 
                      title: '',
                      listItems: []
}

  beforeEach(() => {
    wrapper = shallow(<NoteForm postNote={mockFunc} editNote={mockFunc}  />);
  });

  it("should have a proper default state", () => {
    expect(wrapper.state()).toEqual(mockState);
  });

  it("should match the snapshot with all data passed in correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  describe('addListItem', () => {
    
  })
  it.only("should updateState when AddNote is invoked", () => {
    const mockEvent = {target: {
      parentElement: {
        id: 12
      }
    }}
    
    expect(wrapper.state("listItems")).toHaveLength(0);
   
    expect(wrapper.state("listItems")).toEqual();
  });

  it("should test if state is changed when componentDidMount is invoked", () => {
    wrapper.instance().componentDidMount();
    expect(wrapper.state("listItems")).toHaveLength(1);
  });

  it("should test if state is changed when deleteNote is invoked", () => {
    wrapper.setState({ listItems: [{ id: 12 }] });
    wrapper.instance().deleteNote({
      target: {
        parentElement: {
          id: 12
        }
      }
    });
    expect(wrapper.state("listItems")).toHaveLength(0);
  });

  it("should toggle the checkbox when toggleCheckBox is invoked", () => {
    wrapper.setState({
      listItems: [{ id: 12, checked: false, text: "Test Text" }]
    });
    wrapper.instance().toggleCheckBox({
      target: {
        parentElement: {
          id: 12
        }
      }
    });
    expect(wrapper.state("listItems")).toEqual([
      { id: 12, checked: true, text: "Test Text" }
    ]);
  });

  it("should edit text property when editNoteText is invoked", () => {
    wrapper.setState({
      listItems: [{ id: 12, checked: false, text: "Test Text" }]
    });
    wrapper.instance().editNoteText({
      target: {
        value: "Test2",
        parentElement: {
          id: 12
        }
      }
    });
    expect(wrapper.state("listItems")).toEqual([
      { id: 12, checked: false, text: "Test2" }
    ]);
  });

  it("should edit title property when editTitle is invoked", () => {
    wrapper.setState({ title: "Tester" });
    wrapper.instance().editTitle({
      target: {
        value: "It passes",
        parentElement: {
          id: 12
        }
      }
    });
    expect(wrapper.state("title")).toEqual("It passes");
  });
});
