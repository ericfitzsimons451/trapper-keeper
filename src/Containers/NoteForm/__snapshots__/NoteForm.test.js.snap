// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteForm should match the snapshot with all data passed in correctly 1`] = `
<div
  className="modal"
>
  <form
    className="note-form"
  >
    <input
      className="title-input"
      name="title"
      onChange={[Function]}
      placeholder="title"
      value=""
    />
    <div
      className="text"
      id={1554907871998}
      key="1554907871998"
    >
      <div
        className="add"
      />
      <input
        name="body"
        onChange={[Function]}
        placeholder="take a note"
        value=""
      />
    </div>
    <nav>
      <button
        className="submit-btn"
        onClick={[Function]}
      >
        Save Note
      </button>
    </nav>
  </form>
</div>
`;
